File: go/apskit/adapter.go
package apskit

import (
	"bufio"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"os"
	"os/exec"
	"strconv"
	"strings"
	"time"
)

// JobRequest is one mutator->adapter message.
type JobRequest struct {
	ID           string `json:"id"`
	FeatureJSON  string `json:"feature_json"`
	GeneratedDir string `json:"generated_dir,omitempty"`
	WorkDir      string `json:"work_dir,omitempty"`
	Timeout      string `json:"timeout,omitempty"`
}

// JobResponse is one adapter->mutator message.
type JobResponse struct {
	ID       string `json:"id"`
	Outcome  string `json:"outcome"`
	Output   string `json:"output"`
	Error    string `json:"error"`
	Duration int64  `json:"duration"`
}

func parseTimeout(s string) time.Duration {
	s = strings.TrimSpace(s)
	if s == "" {
		return 0
	}
	switch {
	case strings.HasSuffix(s, "ms"):
		if v, err := strconv.ParseFloat(strings.TrimSuffix(s, "ms"), 64); err == nil {
			return time.Duration(v * float64(time.Millisecond))
		}
	case strings.HasSuffix(s, "s"):
		if v, err := strconv.ParseFloat(strings.TrimSuffix(s, "s"), 64); err == nil {
			return time.Duration(v * float64(time.Second))
		}
	case strings.HasSuffix(s, "m"):
		if v, err := strconv.ParseFloat(strings.TrimSuffix(s, "m"), 64); err == nil {
			return time.Duration(v * float64(time.Minute))
		}
	default:
		if v, err := strconv.ParseFloat(s, 64); err == nil {
			return time.Duration(v * float64(time.Second))
		}
	}
	return 0
}

func classifyExit(code int, killed bool) string {
	if killed {
		return "infrastructure_error"
	}
	if code == 0 {
		return "test_success"
	}
	if code == 1 {
		return "test_failure"
	}
	return "infrastructure_error"
}

// AdapterOptions configures the runner adapter loop.
type AdapterOptions struct {
	Command []string
	Cwd     string
	Stdin   io.Reader
	Stdout  io.Writer
	Stderr  io.Writer
}

func runOne(job JobRequest, opts AdapterOptions) JobResponse {
	timeout := parseTimeout(job.Timeout)
	ctx := context.Background()
	if timeout > 0 {
		var cancel context.CancelFunc
		ctx, cancel = context.WithTimeout(ctx, timeout)
		defer cancel()
	}
	cmd := exec.CommandContext(ctx, opts.Command[0], opts.Command[1:]...)
	cmd.Dir = opts.Cwd
	cmd.Env = append(
		append([]string(nil), os.Environ()...),
		"APS_IR_PATH="+job.FeatureJSON,
		"APS_GENERATED_DIR="+job.GeneratedDir,
		"APS_WORK_DIR="+job.WorkDir,
	)
	start := time.Now()
	stdout, stderr := &strings.Builder{}, &strings.Builder{}
	cmd.Stdout, cmd.Stderr = stdout, stderr
	err := cmd.Run()
	elapsed := time.Since(start).Nanoseconds()
	exitCode := 0
	killed := false
	if err != nil {
		if exitErr, ok := err.(*exec.ExitError); ok {
			exitCode = exitErr.ExitCode()
		} else {
			killed = true
		}
	}
	if ctx.Err() == context.DeadlineExceeded {
		killed = true
	}
	resp := JobResponse{
		ID:       job.ID,
		Outcome:  classifyExit(exitCode, killed),
		Output:   stdout.String(),
		Error:    stderr.String(),
		Duration: elapsed,
	}
	if killed {
		resp.Error = appendError(resp.Error, "infrastructure: process killed or timed out")
	}
	return resp
}

func appendError(existing, msg string) string {
	if existing == "" {
		return msg
	}
	return existing + "\n" + msg
}

// Serve reads NDJSON job requests from opts.Stdin and writes NDJSON responses
// to opts.Stdout until EOF.
func Serve(opts AdapterOptions) error {
	if len(opts.Command) == 0 {
		return fmt.Errorf("Serve: empty command")
	}
	in := bufio.NewScanner(opts.Stdin)
	in.Buffer(make([]byte, 1<<20), 1<<24)
	for in.Scan() {
		line := strings.TrimSpace(in.Text())
		if line == "" {
			continue
		}
		var job JobRequest
		if err := json.Unmarshal([]byte(line), &job); err != nil {
			fmt.Fprintf(opts.Stderr, "aps-adapter: bad job line: %v\n", err)
			continue
		}
		resp := runOne(job, opts)
		out, err := json.Marshal(resp)
		if err != nil {
			fmt.Fprintf(opts.Stderr, "aps-adapter: encode response: %v\n", err)
			continue
		}
		opts.Stdout.Write(out)
		opts.Stdout.Write([]byte("\n"))
	}
	return in.Err()
}

// mutate4go-manifest-begin
// {"version":1,"tested_at":"2026-05-28T18:16:05+02:00","module_hash":"d916442158a0b8329ea98b95646e0892ae4920dcd21f0751c11e7d9a45231375","functions":[{"id":"func/parseTimeout","name":"parseTimeout","line":34,"end_line":58,"hash":"a1dee0aa7d70256f2871809f0c72ae3a3e2332a1c9bd066d8bccb127c6a49201"},{"id":"func/classifyExit","name":"classifyExit","line":60,"end_line":71,"hash":"e883318c180250f7c80709370d81a48200e7f95b170c68ba192b26f1fa82d9d9"},{"id":"func/runOne","name":"runOne","line":82,"end_line":126,"hash":"dfb01cde910c13d69431145107c91f86e5137ed8ba3b2c2acc09942e2c9d6762"},{"id":"func/appendError","name":"appendError","line":128,"end_line":133,"hash":"0e408434d6860c110146560623053b4b5d11ab9d6cf25b2975e05f0f447a6de2"},{"id":"func/Serve","name":"Serve","line":137,"end_line":163,"hash":"4d273fcb51a1f715fd630d013f4b985d083cb582600b5484bf8147ebfc74e658"}]}
// mutate4go-manifest-end

File: rust/aps-kit/src/adapter.rs
use std::io::{BufRead, Write};
use std::path::PathBuf;
use std::process::{Command, Stdio};
use std::time::{Duration, Instant};

use serde::{Deserialize, Serialize};

#[derive(Debug, Deserialize)]
pub struct JobRequest {
    pub id: String,
    pub feature_json: String,
    #[serde(default)]
    pub generated_dir: String,
    #[serde(default)]
    pub work_dir: String,
    #[serde(default)]
    pub timeout: String,
}

#[derive(Debug, Serialize)]
pub struct JobResponse {
    pub id: String,
    pub outcome: String,
    pub output: String,
    pub error: String,
    pub duration: u128,
}

fn parse_timeout(s: &str) -> Option<Duration> {
    let trimmed = s.trim();
    if trimmed.is_empty() {
        return None;
    }
    if let Some(rest) = trimmed.strip_suffix("ms") {
        return rest.parse::<f64>().ok().map(Duration::from_secs_f64).map(|d| d / 1000);
    }
    if let Some(rest) = trimmed.strip_suffix('s') {
        return rest.parse::<f64>().ok().map(Duration::from_secs_f64);
    }
    if let Some(rest) = trimmed.strip_suffix('m') {
        return rest.parse::<f64>().ok().map(|v| Duration::from_secs_f64(v * 60.0));
    }
    trimmed.parse::<f64>().ok().map(Duration::from_secs_f64)
}

fn classify(code: Option<i32>, killed: bool) -> &'static str {
    if killed {
        return "infrastructure_error";
    }
    match code {
        Some(0) => "test_success",
        Some(1) => "test_failure",
        _ => "infrastructure_error",
    }
}

pub struct AdapterOptions {
    pub command: Vec<String>,
    pub cwd: Option<PathBuf>,
}

fn run_one(job: &JobRequest, opts: &AdapterOptions) -> JobResponse {
    let timeout = parse_timeout(&job.timeout);
    let start = Instant::now();
    let mut command = Command::new(&opts.command[0]);
    command
        .args(&opts.command[1..])
        .env("APS_IR_PATH", &job.feature_json)
        .env("APS_GENERATED_DIR", &job.generated_dir)
        .env("APS_WORK_DIR", &job.work_dir)
        .stdout(Stdio::piped())
        .stderr(Stdio::piped());
    if let Some(cwd) = &opts.cwd {
        command.current_dir(cwd);
    }
    let mut child = match command.spawn() {
        Ok(c) => c,
        Err(e) => {
            return JobResponse {
                id: job.id.clone(),
                outcome: "infrastructure_error".into(),
                output: String::new(),
                error: format!("spawn failed: {e}"),
                duration: start.elapsed().as_nanos(),
            };
        }
    };

    let status_result = if let Some(timeout) = timeout {
        wait_with_timeout(&mut child, timeout)
    } else {
        child.wait().map(|s| (s, false))
    };

    let mut output = String::new();
    let mut error = String::new();
    if let Some(mut stdout) = child.stdout.take() {
        use std::io::Read;
        let _ = stdout.read_to_string(&mut output);
    }
    if let Some(mut stderr) = child.stderr.take() {
        use std::io::Read;
        let _ = stderr.read_to_string(&mut error);
    }

    match status_result {
        Ok((status, killed)) => JobResponse {
            id: job.id.clone(),
            outcome: classify(status.code(), killed).into(),
            output,
            error,
            duration: start.elapsed().as_nanos(),
        },
        Err(e) => JobResponse {
            id: job.id.clone(),
            outcome: "infrastructure_error".into(),
            output,
            error: format!("{error}\nwait: {e}"),
            duration: start.elapsed().as_nanos(),
        },
    }
}

fn wait_with_timeout(
    child: &mut std::process::Child,
    timeout: Duration,
) -> std::io::Result<(std::process::ExitStatus, bool)> {
    let start = Instant::now();
    loop {
        match child.try_wait()? {
            Some(status) => return Ok((status, false)),
            None => {
                if start.elapsed() >= timeout {
                    let _ = child.kill();
                    let status = child.wait()?;
                    return Ok((status, true));
                }
                std::thread::sleep(Duration::from_millis(50));
            }
        }
    }
}

/// Loops reading NDJSON job requests from stdin and writing NDJSON responses to
/// stdout. Diagnostics go to stderr.
pub fn serve(opts: AdapterOptions) -> Result<(), String> {
    if opts.command.is_empty() {
        return Err("empty test command".into());
    }
    let stdin = std::io::stdin();
    let stdout = std::io::stdout();
    let mut out = stdout.lock();
    for line in stdin.lock().lines() {
        let line = line.map_err(|e| e.to_string())?;
        let trimmed = line.trim();
        if trimmed.is_empty() {
            continue;
        }
        let job: JobRequest = match serde_json::from_str(trimmed) {
            Ok(j) => j,
            Err(e) => {
                eprintln!("aps-adapter: bad job line: {e}");
                continue;
            }
        };
        let resp = run_one(&job, &opts);
        let encoded = serde_json::to_string(&resp).map_err(|e| e.to_string())?;
        writeln!(out, "{encoded}").map_err(|e| e.to_string())?;
        out.flush().map_err(|e| e.to_string())?;
    }
    Ok(())
}
